### PR TITLE
[Enhancement] Add PushDownTopNBelowUnionRule to support push down topn below union all 

### PR DESCRIPTION
**Pull request by LiShuMing**
- Add `PushDownTopNBelowUnionRule ` to push topn below UnionAll operator only when
  - topn without  offset
  - topn's limit < `cboPushDownTopNLimit` (1000 by default)

- you can change `cbo_push_down_topn_limit` to decide whether to use it.
  - set cbo_push_down_topn_limit = 0, close this feature

This is very useful if union's child has many output and can be short-cut by topn and topn runtime filter.

TODO:
- we may only push down the 'topn' when the ratio of child output row count to the 'topn' limit  is greater than the threshold to avoid 'topn' overhead, but the stats are not very accurate.

## What type of PR is this:

- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 109 / 112 (97.32%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowUnionRule.java|57|60|95.00%|[58, 65, 70]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java|49|49|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 113 / 116 (97.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowUnionRule.java|62|65|95.38%|[59, 66, 71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java|48|48|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 113 / 116 (97.41%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowUnionRule.java|62|65|95.38%|[59, 66, 71]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java|48|48|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Comment by wanpengfei-git** [comment]
#### [FE Incremental Coverage Report]

:heart_eyes: **pass** : 115 / 117 (98.29%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowUnionRule.java|63|65|96.92%|[59, 66]|
|:large_blue_circle:|com/starrocks/sql/optimizer/Optimizer.java|2|2|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/RuleType.java|1|1|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java|49|49|100.00%|[]|

**Comment by wanpengfei-git** [comment]
#### [BE Incremental Coverage Report]


:heart_eyes: **pass** : 0 / 0 (0%)

**Review by murphyatwork — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java [hunk 1]
```diff
@@ -102,8 +105,7 @@ public List<OptExpression> transform(OptExpression input, OptimizerContext conte
                 .setOrderByElements(topn.getOrderByElements())
                 .setLimit(topn.getLimit())
                 .setTopNType(topn.getTopNType())
-                .setSortPhase(topn.getSortPhase())
-                .setIsSplit(false)
```
Comment (kangkaisen) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java:106]: why remove spilt false flag?

**Review by packy92 — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java (hunk 1 above)
Comment (LiShuMing) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/transformation/PushDownTopNBelowOuterJoinRule.java:106]: `false ` is the default value. no need add it.

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.2

**Pull request merged**